### DISCUSSION TITLE
Identifying a pin

### DISCUSSION THREAD

**Discussion by dcesari** [micropython/micropython#9770] [Using MicroPython] [answered]
Hello to everybody, sorry for the naive first question, I can't figure out how to reliably identify the calling pin into a generic irq callback function. The sources and some tests/examples suggest to use `p.id`, but that attribute seems not to be available at least on esp32 port.
See e.g.:
```
p = Pin(3, Pin.IN)
p.__class__.__dict__
{'init': <function>, 'value': <function>, 'off': <function>, 'on': <function>, 'irq': <function>, 'IN': 1, 'OUT': 3, 'OPEN_DRAIN': 7, 'PULL_UP': 2, 'PULL_DOWN': 1, 'IRQ_RISING': 1, 'IRQ_FALLING': 2, 'WAKE_LOW': 4, 'WAKE_HIGH': 5, 'DRIVE_0': 0, 'DRIVE_1': 1, 'DRIVE_2': 2, 'DRIVE_3': 3}
```
Any kind of unique identifier in principle would be acceptable. Actually there is this possibility:
```
str(p)
'Pin(3)'
```
but is seems quite awkward to use the object string representation (and I am not sure it is allowed in a callback), is there a better portable way to do that?

Thanks, Davide

**Comment by rkompass** [comment, marked as answer]
Apparently not.
Something like
```
def __init__(self, pin, ........):
    self.pin_id = int(str(pin).split(", ")[0][4:])
```
seems to be custom.

**Comment by dcesari** [comment]
Thanks

While investigating further, I found this example, see https://github.com/micropython/micropython/discussions/9433#discussion-4423159 `class IR_GET` at the end of the question, where an instance function ` _cb_pin` is used as irq callback, I guess the second argument `line` after `self` becomes associated with the single Pin instance argument of the callback; if this approach is correct, then the problem is solved, any extra information needed to identify the pin can be carried over by the class instance.

**Reply by rkompass** [reply]
You are right. I even used that before and now see, I didn't read your question carefully and missed that it was about identifying the pin in an ISR. I think this functionality is even documented.

**Comment by dcesari** [comment]
Just to complete the discussion, for future memory, this example should do the job, the IRQ callback has access to the full class instance, including any arbitrary information:

```
from machine import Pin
import time

class PinWithId:
    last = -1
    def __init__(self, id):
        self.pin = Pin(id, mode=Pin.IN, pull=Pin.PULL_UP)
        self.id = id
        self.pin.irq(handler=self._irqcb, trigger = Pin.IRQ_FALLING)
        self.count = 0

    def _irqcb(self, pin):
        self.count+=1
        PinWithId.last = self.id


p1 = PinWithId(2)
p2 = PinWithId(5)
while True:
    print (PinWithId.last, p1.count, p2.count)
    time.sleep(1.0)
```
